### PR TITLE
Bug 2005902: Add support for storage class selection for mcg standalone

### PR DESCRIPTION
**Pull request by afreen23**
- implement https://issues.redhat.com/browse/RHSTOR-2332
 - fixes https://bugzilla.redhat.com/show_bug.cgi?id=2005902
Screenshot from 2021-12-08 03-39-07
Screenshot from 2021-12-08 03-38-56
Screenshot from 2021-12-08 03-39-45
Screenshot from 2021-12-08 03-39-23
Screenshot from 2021-12-08 03-39-14
Screenshot from 2021-12-08 04-00-13
Screenshot from 2021-12-08 04-10-03

### CODE REVIEW

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/backing-storage-step.tsx
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/backing-storage-step.tsx:157]: this condition make sure , mcg standalone is not present RHCS but only IBM.
In 4.9 mcg standalone was blocked for any external storage

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/backing-storage-step.tsx
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/backing-storage-step.tsx:231]: Moved all this logic to ''backingStorage/setType' action

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/backing-storage-step.tsx
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/backing-storage-step.tsx:239]: wrapping under form group , had some of its default css, hence added a class to fix that

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/capacity-and-nodes-step/capacity-and-nodes-step.tsx
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/capacity-and-nodes-step/capacity-and-nodes-step.tsx:89]: The convention is to add "set" prefix, just fixed that here and below

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/create-local-volume-set-step/body.tsx
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/create-local-volume-set-step/body.tsx:72]: For mcg default volume mode - filesystem,
For full deployment, volume mode - block

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/create-local-volume-set-step/create-local-volume-set-step.tsx
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/create-local-volume-set-step/create-local-volume-set-step.tsx:229]: For mcg default volume mode - filesystem,
For full deployment, volume mode - block

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/review-and-create-step/review-and-create-step.tsx
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/review-and-create-step/review-and-create-step.tsx:23]: Just thought h3 looks better, let me know if any objections

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/review-and-create-step/review-and-create-step.tsx
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/review-and-create-step/review-and-create-step.tsx:81]:
```suggestion
        {!isStandaloneExternal && !iRhcs &&(
```
Will fix with other reviews

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/security-and-network-step/encryption.tsx
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/security-and-network-step/encryption.tsx:147]: just initializing the state for encrytion with MCG.
Cluster wide encryption need to be true by default for MCG

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/security-and-network-step/encryption.tsx
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/security-and-network-step/encryption.tsx:162]: This hook updates validations which is not required for MCG.
Since with MCG only, storage class encryption not supported

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/footer.tsx
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/footer.tsx:54]:
```suggestion
      return !!sc.name && !!deployment;
```

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/review-and-create-step/review-and-create-step.tsx
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/review-and-create-step/review-and-create-step.tsx]: Reverted

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-steps.tsx
Comment (vbnrh) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-steps.tsx:199]:
```suggestion
```

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx:36]:
```suggestion
      // 'value' on SelectProps['onSelect'] is string hence does not match with payload of type "DeploymentType"
```

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/create-local-volume-set-step/body.tsx
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/create-local-volume-set-step/body.tsx:74]: Add a proper comment on why you are disabling this rule

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/backing-storage-step.tsx
Comment (vbnrh) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/backing-storage-step.tsx:227]: Add data-test id to all radios

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts:155]:
```suggestion
```
`newState` is a clone of `prevState` why are we passing two exact values? Can you not use `Object.assign` to update values.

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts:125]:
```suggestion
```
I don't see this required at all. Use `Object.assign{}` to override certain properties.

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts:196]: Since you're adding types. Please add types to the Line 195 as well.

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts:175]: Why is this logic required? Not having this code should keep the logic intact.

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts:147]: Is this logic really required?

**Review by bipuladh — CHANGES_REQUESTED** [review]

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx
Comment (vbnrh) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx:23]: Screenshot from 2021-12-08 19-24-03
I am unable to see this description for full deployment. Please check

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx
Comment (vbnrh) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx:19]: we need to sentence case ` Multicloud Object Gateway ` right ?

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts:147]: Yes its required and I added comment

**Review by vbnrh — CHANGES_REQUESTED** [review]
Tested out. It is working great .

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts:175]: yes required added comment

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/create-local-volume-set-step/body.tsx
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/create-local-volume-set-step/body.tsx:74]: The deps array explains it - it should be rendered once

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx:23]: yeah I added later , you may check on latest

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx:19]: This was given by doc "MultiCloud Object Gateway"

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts:196]: L195 has types `export const reducer: WizardReducer` :D

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts:125]: Yeah, I just though I wont confuse for readability, but tahnks for second opinion

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts:155]: same as below

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/create-local-volume-set-step/body.tsx
Comment (bipuladh) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/create-local-volume-set-step/body.tsx:74]: sorry it's not too clear. How does adding dispatch to the deps array rerender.

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx:36]: is my english wrong here ?

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts
Comment (bipuladh) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts:147]: it says it's required for step 1 pages only why can't we reset the state for all cases. What difference would it make? Could you please elaborate the comment was not clear enough hence I have added this comment.

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/create-local-volume-set-step/body.tsx
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/create-local-volume-set-step/body.tsx:74]: Plus see comments above L72

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-steps.tsx
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-steps.tsx:199]: ack

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx
Comment (vbnrh) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx:19]: We should change it for consistency (either Full Deployment or Multicloud object gateway. Looks weird to me to have two different casing for two distinct items in a dropdown

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/backing-storage-step.tsx
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/backing-storage-step.tsx:227]: can we add them once we update tests ?

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx:19]: Okay got it.
Full deployment - is a general sentence
Multicloud Object Gateway - is the product name, so its fine.

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/backing-storage-step.tsx
Comment (bipuladh) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/backing-storage-step.tsx:227]: Please add along with tests. We don't want ids which are not being referenced anywhere.

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx
Comment (bipuladh) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx:36]: sort of.

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts
Comment (bipuladh) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts:188]: sorry my bad

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts]: It saying not req for step 1 =
`Its not required if the user has not visited more than first step` -> more than a single step.
```
  /*
   * Wizard state should be reset when a new backing storage type is selected
   * in order to avoid stale state collisions since each backing storage type
   * has its own supported variables. e.g if arbiter was selected and not
   * deselected before changing the backing storage type then storage cluster spec
   * will mark the arbiter option as enabled.
   *
   * Its not required if the user has not visited more than first step.
   */
   ```

   Its to avoid stale states, every mode of deployment sets some options in global state, if that mode f deployment is dropped it should be reset , other wise previous global state will be overwritten.
  Visually you will have steps visited in wizard nav even though user did not actually went through those.
  And code wise you can have kms enabled , arbiter network options enabled in diff modes where they are not supported and you will have a bad stoarge cluster yaml. So a couple of bad effects if not reset.
  This is all in 4.9 too , I just isolated it here so it will be easy to follow .

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/backing-storage-step/select-deployment.tsx]: changed

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/create-local-volume-set-step/body.tsx
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/create-local-volume-set-step/body.tsx:75]: added comment as well , check now!

Comment (bipuladh) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/create-storage-system-steps/create-local-volume-set-step/body.tsx:75]: If it's only required to run once. Is `useEffect` really required?

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/create-storage-system/reducer.ts]: Therefore for first step , its not needed to reset whole global state, since user has not visited more steps.
just doing when there are steps visited and then deselection happening

**Review by bipuladh — APPROVED** [review]

**Comment by afreen23** [comment]
/hold

Holding the merge due to https://bugzilla.redhat.com/show_bug.cgi?id=2031023 in backend.
If we merge this then it will break 4.9 UI too and in case issue not resolved will make 4.10 code to be reverted.
/ cc @iamniting @tamireran @nthomas-redhat @jarrpa

**Comment by iamniting** [comment]
The backend PR https://github.com/red-hat-storage/odf-operator/pull/152 is merged in the main branch. We are in discussion to backport it to 4.9.z.

**Pull request merged**
